### PR TITLE
[Layer] Add SequenceMeta, SequencePool and init pipeline parrallel function.

### PR DESCRIPTION
**Pull request by changqi1**

### CODE REVIEW

File: src/models/common_decoder.h [hunk 1]
```diff
@@ -341,15 +343,46 @@ class CommonDecoder : public AbstractDecoder {
         if (ctx->ppSize > 1 && ctx->ppRank > 0) {
             int curr_world_rank = ctx->ppRank * ctx->tpSize + ctx->tpRank;
             int prev_world_rank = (ctx->ppRank - 1) * ctx->tpSize + ctx->tpRank;
-            int count = batchSize * inputSeqLen * ctx->hiddenSize;
+            int count = batchSize * inputSeqLen * hiddenSize;
+            TimeLine t("Decoder.MPI_Recv." + std::to_string(ctx->promptID));
+            printf("%d: Decoder.MPI_Recv.SyncStart\n", ctx->ppRank);
+            fflush(stdout);
+            int32_t promptID;
+            MPI_Recv(&promptID, 1, MPI_INT32_T, prev_world_rank, curr_world_rank, MPI_COMM_WORLD, MPI_STATUS_IGNORE);
```
Comment (Duyi-Wang) [inline src/models/common_decoder.h:351]: input prompt has been synced in model.input().

File: src/common/sample_info.h [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+        return instance;
+    }
+
+    bool empty() { return queue.empty(); }
+
+    int32_t size() { return queue.size(); }
+
+    bool isFull() {
+        bool full = false;
+        if (this->size() >= 4) full = true;
```
[the remark this answers is not in the document]
Comment (Duyi-Wang) [reply src/common/sample_info.h:155]: 环境变量？我开始误以为是pipline中有多少张卡的数量。

File: src/common/sample_info.h [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+private:
+    int32_t sampleID;
+    int32_t inputSeqLen;
+    int32_t hiddenSize;
+    int32_t pastSeqLen;
+    std::vector<int32_t> inputTokens;
+    std::vector<int32_t> pastTokens; // generated tokens
+
+    // Indicates whether the sample is in the prefill phase
+    bool bePrefill;
```
[the remark this answers is not in the document]
Comment (Duyi-Wang) [reply src/common/sample_info.h:93]: 改成isPrompt? prefill不太好理解。 vllm就用的这个变量名。

File: src/common/sample_info.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (changqi1) [reply src/common/sample_info.h]: 这个值也和几张卡有关的

**Comment by changqi1** [comment]
Done.

models/models.cpp
@@ -85,6 +85,13 @@ void Model::input(std::vector<int32_t> &inputIds_, int batchSize_) {
     inputIds.resize(dims[1]);
     if (decoder->getRank() == 0) { inputIds = inputIds_; }
     messenger.broadcast(inputIds.data(), dims[1]);
+
+    if (this->isMaster()) {
+        for (int i = 0; i < 2; ++i) {
+            int sequenceID = InputQueue::getInstance().createSequenceID();
+            InputQueue::getInstance().push(SequencePool::getInstance().createMeta(sequenceID, seqLen, inputIds));
+        }
+    }
 }

File: src/common/sequence.h [hunk 4]
```diff
@@ -0,0 +1,275 @@
...
+    // For next tokens
+    void addNextToken(int32_t token) { nextTokens.push_back(token); }
+
+    int32_t getLatestToken() const { return nextTokens.back(); }
+
+    const int32_t *getTotalTokens() const { return nextTokens.data(); }
+
+    int32_t getStep() const { return step; }
+
+    void setStep(int32_t _step) { step = _step; }
```
Comment (pujiang2018) [inline src/common/sequence.h:79]: Is it better to design function like below to merge addNextToken and setStep:
void stepForward(int id)
(and in future, the id could be a vector)
inside that function, we can manage step and other changed members.

File: src/common/transformer_ctx.h [hunk 5]
```diff
@@ -64,6 +64,7 @@ struct DecoderContext {
     int inputSeqLen;
     // For custom usage
     int reserved1;
+    int sequenceID;
```
Comment (pujiang2018) [inline src/common/transformer_ctx.h:67]: why do we need it here?

Comment (changqi1) [reply src/common/transformer_ctx.h:67]: Searchers don't know which sequenceID is executing, and need to transmit the sequenceID to Searchers.
If don't put it in DecoderContext, we could put it in the returned result in CommonDecoder.forward().

Comment (changqi1) [reply src/common/transformer_ctx.h:69]: Done

File: src/common/sequence.h (hunk 4 above)
Comment (changqi1) [reply src/common/sequence.h:99]: Done

File: src/common/sequence.h [hunk 6]
```diff
@@ -0,0 +1,286 @@
...
+    std::vector<SequenceMeta *> sequences;
+};
+
+//    SequencePool
+//    ┌──────┬──────┬──────┐
+//    │      │      │  ◄───┼──┬─ SequenceMeta
+//    ├──────┼──────┼──────┤  │
+//    │      │      │  ◄───┼──┘
+//    └──────┴──────┴──────┘
+class SequencePool {
```
Comment (Duyi-Wang) [inline src/common/sequence.h:142]: Make SequenceGroup as the base unit of Sequencepool.

Comment (Duyi-Wang) [reply src/common/sequence.h:142]: Just make it work when num_beam=1, I will finish it for beam search in the future.

**Pull request closed without merging**